niworkflows/interfaces/bids.py

```diff
@@ -392,110 +407,111 @@ class DerivativesDataSink(SimpleInterface):
+        for k in self._allowed_entities.intersection(list(inputs.keys())):
+            setattr(self.inputs, k, inputs[k])
 
-        self._results['out_file'] = []
         if out_path_base:
             self.out_path_base = out_path_base
 
     def _run_interface(self, runtime):
-        if isdefined(self.inputs.meta_dict):
-            meta = self.inputs.meta_dict
-            # inputs passed in construction take priority
-            meta.update(self._metadata)
-            self._metadata = meta
-
-        src_fname, _ = _splitext(self.inputs.source_file)
-        src_fname, dtype = src_fname.rsplit('_', 1)
-        _, ext = _splitext(self.inputs.in_file[0])
-        if self.inputs.compress is True and not ext.endswith('.gz'):
-            ext += '.gz'
-        elif self.inputs.compress is False and ext.endswith('.gz'):
-            ext = ext[:-3]
-
-        m = BIDS_NAME.search(src_fname)
-
-        mod = Path(self.inputs.source_file).parent.name
-
+        # Ready the output folder
         base_directory = runtime.cwd
         if isdefined(self.inputs.base_directory):
             base_directory = self.inputs.base_directory
+        base_directory = Path(base_directory).absolute()
+        out_path = base_directory / self.out_path_base
+        out_path.mkdir(exist_ok=True, parents=True)
 
-        base_directory = Path(base_directory).resolve()
-        out_path = base_directory / self.out_path_base / \
-            '{subject_id}'.format(**m.groupdict())
-
-        if m.groupdict().get('session_id') is not None:
-            out_path = out_path / '{session_id}'.format(**m.groupdict())
+        # Ensure we have a list
+        in_file = listify(self.inputs.in_file)
 
-        out_path = out_path / '{}'.format(mod)
-        out_path.mkdir(exist_ok=True, parents=True)
-        base_fname = str(out_path / src_fname)
+        # Read in the dictionary of metadata
+        if isdefined(self.inputs.meta_dict):
+            meta = self.inputs.meta_dict
+            # inputs passed in construction take priority
+            meta.update(self._metadata)
+            self._metadata = meta
 
-        allowed_entities = {}
+        # Initialize entities with those from the source file.
+        out_entities = parse_file_entities(self.inputs.source_file)
+        for drop_entity in listify(self.inputs.dismiss_entities or []):
+            out_entities.pop(drop_entity, None)
+
+        # Override extension with that of the input file(s)
+        out_entities["extension"] = [
+            "".join(Path(orig_file).suffixes).lstrip(".") for orig_file in in_file
+        ]
+
+        compress = listify(self.inputs.compress) or [None]
+        if len(compress) == 1:
+            compress = compress * len(in_file)
+        for i, ext in enumerate(out_entities["extension"]):
+            if compress[i] is not None:
+                ext = ext.rstrip(".gz")
+                out_entities["extension"][i] = f"{ext}{'.gz' * bool(compress[i])}"
```

Comment:
Finally went with
```Python 
out_entities["extension"][i] = f"{ext}.gz" if compress[i] else ext
```
because it is significantly faster than any other option, and also more readable after all (as you point out elsewhere).